### PR TITLE
Merge branches storage

### PR DESCRIPTION
**Pull request by diega**

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 2 issues

1. ![MAJOR][MAJOR] [ContractDetailsImpl.java#L389](https://github.com/rsksmart/rskj/blob/4eb8d3ca522141c64078af524e25015fb98c87cd/rskj-core/src/main/java/co/rsk/db/ContractDetailsImpl.java#L389): Inconsistent synchronization of co.rsk.db.ContractDetailsImpl.trie; locked 78% of time [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3AIS2_INCONSISTENT_SYNC)
1. ![MAJOR][MAJOR] [DefaultConfig.java#L75](https://github.com/rsksmart/rskj/blob/4eb8d3ca522141c64078af524e25015fb98c87cd/rskj-core/src/main/java/org/ethereum/config/DefaultConfig.java#L75): java/io/File.<init>(Ljava/lang/String;)V reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: rskj-core/src/main/java/co/rsk/db/ContractDetailsImpl.java [hunk 1]
```diff
@@ -364,14 +360,18 @@ public synchronized ContractDetails getSnapshotTo(byte[] hash) {
...
+                                                              this.trieStoreFactory,
+                                                              this.memoryStorageLimit);
         details.keys = new HashSet<>();
         details.keys.addAll(this.keys);
         details.externalStorage = this.externalStorage;
         details.originalExternalStorage = this.originalExternalStorage;
 
         if (this.externalStorage) {
-            levelDbByName(getDataSourceName(), databaseDir);
+            DataSourcePool.reserve(getDataSourceName());
```
Comment (SergioDemianLerner) [inline rskj-core/src/main/java/co/rsk/db/ContractDetailsImpl.java]: Pregunta: porque el cambio de levelDBByName*( a DataSourcePool.reserve(). Cual es la diferencia?

File: rskj-core/src/main/java/co/rsk/db/ContractDetailsImpl.java [hunk 2]
```diff
@@ -401,7 +407,7 @@ private void removeKey(byte[] key) {
     }
 
     private void checkExternalStorage() {
-        this.externalStorage = (keys.size() > memoryStorageLimit) || this.externalStorage;
+        this.externalStorage = true;
```
Comment (SergioDemianLerner) [inline rskj-core/src/main/java/co/rsk/db/ContractDetailsImpl.java]: Porque decidieron no usar mas el encoding liviano ?

File: rskj-core/src/main/java/co/rsk/db/RepositoryImpl.java [hunk 3]
```diff
@@ -349,17 +357,18 @@ public synchronized void updateBatch(Map<RskAddress, AccountState> stateCache,
...
                     originalContractDetails.setAddress(addr.getBytes());
                     contractDetailsCache.setOriginalContractDetails(originalContractDetails);
                     contractDetailsCache.commit();
                 }
 
                 contractDetails = contractDetailsCache.getOriginalContractDetails();
+                ContractDetails storeContractDetails = detailsDataStore.get(addr);
 
-                updateContractDetails(addr, contractDetails);
+                updateContractDetails(addr, contractDetails, storeContractDetails);
```
Comment (SergioDemianLerner) [inline rskj-core/src/main/java/co/rsk/db/RepositoryImpl.java]: Como justificarían que este es el único lugar posible donde es necesario hacer el merge ?

**Comment by rskops** [comment]
SonarQube analysis reported 2 issues

1. ![MAJOR][MAJOR] [ContractDetailsImpl.java#L389](https://github.com/rsksmart/rskj/blob/a031cb683cf3bd63c3bd8e6ba47de163eb4cec8f/rskj-core/src/main/java/co/rsk/db/ContractDetailsImpl.java#L389): Inconsistent synchronization of co.rsk.db.ContractDetailsImpl.trie; locked 78% of time [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3AIS2_INCONSISTENT_SYNC)
1. ![MAJOR][MAJOR] [DefaultConfig.java#L75](https://github.com/rsksmart/rskj/blob/a031cb683cf3bd63c3bd8e6ba47de163eb4cec8f/rskj-core/src/main/java/org/ethereum/config/DefaultConfig.java#L75): java/io/File.<init>(Ljava/lang/String;)V reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by rskops** [comment]
SonarQube analysis reported 1 issue

1. ![MAJOR][MAJOR] [DefaultConfig.java#L75](https://github.com/rsksmart/rskj/blob/2974626ccb0a15c522895c5a222322fca0ef814c/rskj-core/src/main/java/org/ethereum/config/DefaultConfig.java#L75): java/io/File.<init>(Ljava/lang/String;)V reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**